### PR TITLE
Improve Cell navigation Performance (using rxjs proof of concept)

### PR DESCRIPTION
**Pull request by malonecj**
## Description
This is an attempt to see if its possible to perform cell navigation with keyboard at 60FPS. 
Rather than the `Cell` component being responsible for displaying a selected style, a new `SelectionMask` component which sits at `Canvas` level, is responsible for displaying the selected cell.

## Results
The proof of concept was successful, showing that every frame is rendered at over 60FPS.
image

c11e914ad04e41a1aa3f0000bfe7b0bd

## Tasks
 - [x] Automated benchmark test using Puppeteer
 - [x] SelectionMask component
 - [x] Dynamic calculation of mask dimensions based on idx, rowIdx of cell
 - [x] Connect function to wrap any component, connecting it to a single store and trigger set state, if certain conditions are met
 - [x] Wrap cell selector in React Motion wrapper for smooth cell animation
 - [ ] Scrolling with keyboard - now that selected cells do not automatically focus, the grid no longer scrolls with keyboard (rudimentary proof of concept for scroll down working)

## Further Work
 - [ ] Cell editing - this should be trivial, and is close to working
 - [ ] Frozen columns

 
This pr was based off [a previous pull request](https://github.com/adazzle/react-data-grid/pull/1096), that uses a state container with rxjs to both publish actions, and subscribe to them using a Connect function, similar to how Redux works. Whereas this proof of concept could easily be done, without this wrapper, it made it very simple to add the connected SelectionMask component, as all its necessary props were already in shared state. 

 

## Use of rxjs
 Pros 
- Easily subscribe to state change from any component
- Greater control of what and when we render
- Allows all components to be stateless and simple, with complexity hidden away in state management folder
- No more need for refs - having connected components mean that we will not need to query down the component tree to access props and state of child components

## Cons
 - More overhead - we can import just the parts of the library that we need but is it overkill?
 - Not trivial to understand - however, if we keep this part to a core section or even extract out to another package with a very simple api, that way any component can be connected
Any others?

### CODE REVIEW

File: packages/react-data-grid/src/Cell.js [hunk 1]
```diff
@@ -71,43 +71,21 @@ class Cell extends React.Component {
...
-      || this.props.isCellValueChanging(this.props.value, nextProps.value)
-      || this.props.forceUpdate === true
-      || this.props.className !== nextProps.className
-      || this.props.expandableOptions !== nextProps.expandableOptions
-      || this.hasChangedDependentValues(nextProps)
-      || this.props.column.locked !== nextProps.column.locked;
-    return shouldUpdate;
-  }
-
   onCellClick = (e) => {
```
Comment (malonecj) [inline packages/react-data-grid/src/Cell.js]: using rxjs should allow us to reduce the complexity of our shouldComponentUpdate functions

File: packages/react-data-grid/src/Cell.js [hunk 2]
```diff
@@ -21,9 +21,9 @@ class Cell extends React.Component {
   static propTypes = {
     rowIdx: PropTypes.number.isRequired,
     idx: PropTypes.number.isRequired,
-    selected: PropTypes.shape({
-      idx: PropTypes.number.isRequired
-    }),
+    isSelected: PropTypes.bool,
```
Comment (malonecj) [inline packages/react-data-grid/src/Cell.js]: code no longer necessary now that using SelectionMask

File: packages/react-data-grid/src/ReactDataGrid.js [hunk 3]
```diff
@@ -150,13 +159,7 @@ class ReactDataGrid extends React.Component {
...
     super(props, context);
     let columnMetrics = this.createColumnMetrics();
-    let initialState = {columnMetrics, selectedRows: [], copied: null, expandedRows: [], canFilter: false, columnFilters: {}, sortDirection: null, sortColumn: null, dragged: null, scrollOffset: 0, lastRowIdxUiSelected: -1};
-    if (props.enableCellSelect) {
-      initialState.selected = {rowIdx: 0, idx: 0};
-    } else {
-      initialState.selected = {rowIdx: -1, idx: -1};
-    }
-    this.state = initialState;
+    this.state = {columnMetrics, selectedRows: [], copied: null, expandedRows: [], canFilter: false, columnFilters: {}, sortDirection: null, sortColumn: null, dragged: null, scrollOffset: 0, lastRowIdxUiSelected: -1};
```
Comment (malonecj) [inline packages/react-data-grid/src/ReactDataGrid.js]: Have removed state.selected - this now lives in single store

File: packages/react-data-grid/src/ReactDataGrid.js [hunk 4]
```diff
@@ -331,26 +329,14 @@ class ReactDataGrid extends React.Component {
...
-            this.props.onCellSelected(selected);
-          }
-        });
-      } else if (rowIdx === -1 && idx === -1) {
-        // When it's outside of the grid, set rowIdx anyway
-        this.setState({selected: { idx, rowIdx }});
-      }
+  onSelect = (selected) => {
+    const {selectCell, onCellDeSelected, onCellSelected} = this.props;
+    selectCell(selected);
```
Comment (malonecj) [inline packages/react-data-grid/src/ReactDataGrid.js]: no more setting state on each cell selection, this now triggers an action, and any component can subscribe to state change if necessary

File: packages/react-data-grid/src/connectedComponents/SelectionMask.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+import { Motion, spring } from 'react-motion';
+import { connect } from '../stateManagement/state/RxState';
+import initialState from '../stateManagement/state/initialState';
+
+/* MapStateToProps */
+
+const Mask = ({ selectedPosition }) => {
+  const { idx, rowIdx } = selectedPosition;
+  const [width, height] = [150, 35];
```
Comment (malonecj) [inline packages/react-data-grid/src/connectedComponents/SelectionMask.js]: assuming fixed cell dimensions to calculate position. Need to figure out the optimal way to obtain this info

**Comment by hutber** [comment]
Would love to be able to test this out as well :)

**Pull request closed without merging**